File: src/main/resources/harness/workspace/imports/Formatter.java
package com.example.util;

/**
 * Bottom of the chain — depends on nothing but the JDK.
 *
 * <p>Part of the M15 cross-file fixture. Three files, three different ways of reaching another one:</p>
 *
 * <ul>
 *   <li>{@code Greeter} calls this with <b>no import</b>, because they share a package.</li>
 *   <li>{@code com.example.Main} reaches {@code Greeter} <b>through an import</b>, from another package.</li>
 *   <li>and {@code Main} therefore depends on this one <b>transitively</b>, without naming it at all.</li>
 * </ul>
 *
 * <p><b>Things worth trying here.</b> Rename {@code shout} and watch {@code Greeter} go red on the next
 * analysis without saving anything — the index answers from the open buffer, not the file. Change the
 * return type to {@code int} and watch the error appear in {@code Greeter}, one file away. Delete the
 * method entirely and both other files should still parse, colour and fold; only resolution goes.</p>
 */
public final class Formatter {

    private static final String BANG = "!";

    private Formatter() {
    }

    /** Trims, upper-cases and punctuates. Null-safe, so the fixture has a branch in it. */
    public static String shout(String text) {
        if (text == null || text.isEmpty()) return "";
        return text.trim().toUpperCase() + BANG;
    }

    /** A second member, so completion on {@code Formatter.} has more than one row to show. */
    public static String quiet(String text) {
        return text == null ? "" : text.trim().toLowerCase();
    }
}

File: src/main/resources/harness/workspace/imports/Formatter.js
// Bottom of the chain — imports nothing, and exports without saying so.
//
// There is no `exports.` anywhere in this file. A module that says nothing exports every TOP-LEVEL
// declaration, which is the default because this language already writes its own `import util.Formatter;`
// rather than ES or CommonJS syntax — so requiring Node's `exports.name = …` on the other side would
// leave you writing a Java-shaped import against a Node-shaped export.
//
// It is also better for the editor, and that is not a side effect. `function shout(text)` is a
// DECLARATION: it carries a kind, its parameter names and a precise span, so a hover renders
// `function shout(text)` and Ctrl+B lands on the word. `exports.shout = function (text)` is an assignment
// of an anonymous function to a property — statically that offers a name and very little else.
//
// Part of the M15 S6/S7 fixture. Three files, mirroring the Java one next door:
//
//   * App.js             imports util.Greeter and runs it
//   * util/Greeter.js    imports util.Formatter — a module that itself imports a module
//   * util/Formatter.js  this one, the leaf
//
// Things worth trying here:
//   * Add `function whisper(t) { return t.toLowerCase(); }` — it should appear behind `Formatter.` in
//     Greeter.js on the next analysis, with no export line to write.
//   * Rename `shout` and watch Greeter.js go red, with NO SAVE.
//   * Want something PRIVATE? Add one `exports.shout = shout;` line and the rest of the file stops being
//     exported — an explicit export is taken at its word, which is the only way to hide a helper.

function shout(text) {
    if (!text) {
        return '';
    }
    return String(text).trim().toUpperCase() + '!';
}

function quiet(text) {
    return text ? String(text).trim().toLowerCase() : '';
}
